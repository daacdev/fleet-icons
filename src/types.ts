import React from 'react';
import { StyledComponent } from 'styled-components';

export interface IconProps extends React.SVGAttributes<SVGSVGElement> {
  readonly size?: number | string;
  readonly rotate?: number;
  readonly spinner?: boolean;
  readonly title?: string;
  readonly desc?: string;
}

export type IconComponentType = StyledComponent<
  'svg',
  Record<string, unknown>,
  IconProps,
  never
>;
