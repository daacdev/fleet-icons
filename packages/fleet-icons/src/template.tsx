import * as React from 'react';
import styled, { keyframes, css } from 'styled-components';
import { IconProps, IconComponentType } from './types';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

const animation = css`
  animation: ${rotate} 2s linear infinite;
`;

export const GenerateIcon = (
  children: React.ReactElement[]
): IconComponentType => {
  return styled.svg.attrs<IconProps>(({ title, desc }) => ({
    children: (
      <>
        {title && <title> {title} </title>}
        {desc && <desc> {desc} </desc>}
        {children.map((Element, key) => (
          <Element.type {...Element.props} key={key} />
        ))}
      </>
    ),
    xmlns: 'http://www.w3.org/2000/svg',
    viewBox: '0 0 16 16',
  }))<IconProps>`
    fill: ${({ fill }) => fill || 'currentColor'};
    stroke: ${({ stroke }) => stroke || 'currentColor'};
    stroke-width: ${({ strokeWidth }) => strokeWidth || 0};
    width: ${({ size }) => size || '1em;'};
    height: ${({ size }) => size || '1em;'};
    ${({ spinner }) => spinner && animation}
    ${({ rotate }) => rotate && `transform: rotate(${rotate}deg);`}
  `;
};
