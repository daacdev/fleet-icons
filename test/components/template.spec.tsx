import React from 'react';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';

import { GenerateIcon } from '../../src/template';

const svg_paths_d = [
  'M10.348 7.643c0-1.112.488-1.754 1.318-1.754.682 0 1.139.47 1.187 1.108H14v-.11c-.053-1.187-1.024-2-2.342-2-1.604 0-2.518 1.05-2.518 2.751v.747c0 1.7.905 2.73 2.518 2.73 1.314 0 2.285-.792 2.342-1.939v-.114h-1.147c-.048.615-.497 1.05-1.187 1.05-.839 0-1.318-.62-1.318-1.727v-.742zM4.457 11l1.02-4.184h.045L6.542 11h1.006L9 5.001H7.818l-.82 4.355h-.056L5.97 5.001h-.94l-.972 4.355h-.053l-.827-4.355H2L3.452 11h1.005z',
  'M14 3a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h12zM2 2a2 2 0 0 0-2 2v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V4a2 2 0 0 0-2-2H2z',
];

const Paths = svg_paths_d.map((d, key) => <path key={key} d={d} />);
const Icon = GenerateIcon(Paths);

test('Simple rendering Icon component without props', () => {
  const { container } = render(<Icon />);
  expect(container).toMatchSnapshot();
});

test('Rendering Icon component with spinner prop', () => {
  const { container } = render(<Icon spinner />);
  expect(container).toMatchSnapshot();
});

test('Rendering Icon component with rotate prop', () => {
  const { container } = render(<Icon rotate={45} />);
  expect(container).toMatchSnapshot();
});

test('Rendering Icon component with size prop', () => {
  const { container } = render(<Icon size="45px" />);
  expect(container).toMatchSnapshot();
});

test('Icon component accessibility', () => {
  render(<Icon title="Icon" desc="An example of an Icon icon" />);

  const IconTest = screen.getAllByTitle('Icon');
  expect(IconTest).toMatchSnapshot();
});
