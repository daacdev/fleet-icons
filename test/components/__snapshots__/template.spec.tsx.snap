// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component accessibility 1`] = `
Array [
  .c0 {
  fill: currentColor;
  stroke: currentColor;
  stroke-width: 0;
  width: 1em;
  height: 1em;
}

<svg
    class="c0"
    title="Icon"
    viewBox="0 0 16 16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
       
      Icon
       
    </title>
    <desc>
       
      An example of an Icon icon
       
    </desc>
    <path
      d="M10.348 7.643c0-1.112.488-1.754 1.318-1.754.682 0 1.139.47 1.187 1.108H14v-.11c-.053-1.187-1.024-2-2.342-2-1.604 0-2.518 1.05-2.518 2.751v.747c0 1.7.905 2.73 2.518 2.73 1.314 0 2.285-.792 2.342-1.939v-.114h-1.147c-.048.615-.497 1.05-1.187 1.05-.839 0-1.318-.62-1.318-1.727v-.742zM4.457 11l1.02-4.184h.045L6.542 11h1.006L9 5.001H7.818l-.82 4.355h-.056L5.97 5.001h-.94l-.972 4.355h-.053l-.827-4.355H2L3.452 11h1.005z"
    />
    <path
      d="M14 3a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h12zM2 2a2 2 0 0 0-2 2v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V4a2 2 0 0 0-2-2H2z"
    />
  </svg>,
  <title>
     
    Icon
     
  </title>,
]
`;

exports[`Rendering Icon component with rotate prop 1`] = `
.c0 {
  fill: currentColor;
  stroke: currentColor;
  stroke-width: 0;
  width: 1em;
  height: 1em;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

<div>
  <svg
    class="c0"
    rotate="45"
    viewBox="0 0 16 16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10.348 7.643c0-1.112.488-1.754 1.318-1.754.682 0 1.139.47 1.187 1.108H14v-.11c-.053-1.187-1.024-2-2.342-2-1.604 0-2.518 1.05-2.518 2.751v.747c0 1.7.905 2.73 2.518 2.73 1.314 0 2.285-.792 2.342-1.939v-.114h-1.147c-.048.615-.497 1.05-1.187 1.05-.839 0-1.318-.62-1.318-1.727v-.742zM4.457 11l1.02-4.184h.045L6.542 11h1.006L9 5.001H7.818l-.82 4.355h-.056L5.97 5.001h-.94l-.972 4.355h-.053l-.827-4.355H2L3.452 11h1.005z"
    />
    <path
      d="M14 3a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h12zM2 2a2 2 0 0 0-2 2v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V4a2 2 0 0 0-2-2H2z"
    />
  </svg>
</div>
`;

exports[`Rendering Icon component with size prop 1`] = `
.c0 {
  fill: currentColor;
  stroke: currentColor;
  stroke-width: 0;
  width: 45px;
  height: 45px;
}

<div>
  <svg
    class="c0"
    viewBox="0 0 16 16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10.348 7.643c0-1.112.488-1.754 1.318-1.754.682 0 1.139.47 1.187 1.108H14v-.11c-.053-1.187-1.024-2-2.342-2-1.604 0-2.518 1.05-2.518 2.751v.747c0 1.7.905 2.73 2.518 2.73 1.314 0 2.285-.792 2.342-1.939v-.114h-1.147c-.048.615-.497 1.05-1.187 1.05-.839 0-1.318-.62-1.318-1.727v-.742zM4.457 11l1.02-4.184h.045L6.542 11h1.006L9 5.001H7.818l-.82 4.355h-.056L5.97 5.001h-.94l-.972 4.355h-.053l-.827-4.355H2L3.452 11h1.005z"
    />
    <path
      d="M14 3a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h12zM2 2a2 2 0 0 0-2 2v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V4a2 2 0 0 0-2-2H2z"
    />
  </svg>
</div>
`;

exports[`Rendering Icon component with spinner prop 1`] = `
.c0 {
  fill: currentColor;
  stroke: currentColor;
  stroke-width: 0;
  width: 1em;
  height: 1em;
  -webkit-animation: bKPVWD 2s linear infinite;
  animation: bKPVWD 2s linear infinite;
}

<div>
  <svg
    class="c0"
    viewBox="0 0 16 16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10.348 7.643c0-1.112.488-1.754 1.318-1.754.682 0 1.139.47 1.187 1.108H14v-.11c-.053-1.187-1.024-2-2.342-2-1.604 0-2.518 1.05-2.518 2.751v.747c0 1.7.905 2.73 2.518 2.73 1.314 0 2.285-.792 2.342-1.939v-.114h-1.147c-.048.615-.497 1.05-1.187 1.05-.839 0-1.318-.62-1.318-1.727v-.742zM4.457 11l1.02-4.184h.045L6.542 11h1.006L9 5.001H7.818l-.82 4.355h-.056L5.97 5.001h-.94l-.972 4.355h-.053l-.827-4.355H2L3.452 11h1.005z"
    />
    <path
      d="M14 3a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h12zM2 2a2 2 0 0 0-2 2v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V4a2 2 0 0 0-2-2H2z"
    />
  </svg>
</div>
`;

exports[`Simple rendering Icon component without props 1`] = `
.c0 {
  fill: currentColor;
  stroke: currentColor;
  stroke-width: 0;
  width: 1em;
  height: 1em;
}

<div>
  <svg
    class="c0"
    viewBox="0 0 16 16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10.348 7.643c0-1.112.488-1.754 1.318-1.754.682 0 1.139.47 1.187 1.108H14v-.11c-.053-1.187-1.024-2-2.342-2-1.604 0-2.518 1.05-2.518 2.751v.747c0 1.7.905 2.73 2.518 2.73 1.314 0 2.285-.792 2.342-1.939v-.114h-1.147c-.048.615-.497 1.05-1.187 1.05-.839 0-1.318-.62-1.318-1.727v-.742zM4.457 11l1.02-4.184h.045L6.542 11h1.006L9 5.001H7.818l-.82 4.355h-.056L5.97 5.001h-.94l-.972 4.355h-.053l-.827-4.355H2L3.452 11h1.005z"
    />
    <path
      d="M14 3a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H2a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1h12zM2 2a2 2 0 0 0-2 2v8a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V4a2 2 0 0 0-2-2H2z"
    />
  </svg>
</div>
`;
